python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,726 @@
+        # node_id which needs to be vm.vm and not vm.name
+        with self.lock:
+            category_id = self.get_category()
+            if not category_id:
+                category_id = self.create_category()
+
+            for key, value in tags.items():
+
+                tag = kv_pair_to_vsphere_tag(key, value)
+                tag_id = self.get_tag(tag, category_id)
+                if not tag_id:
+                    tag_id = self.create_node_tag(tag, category_id)
+
+                # If a tag with a key is present on the VM, then remove it
+                # before updating the key with a new value.
+                self.remove_tag_from_vm(key, node_id)
+
+                cli_logger.info(f"Attaching tag {tag} to {node_id}")
+                self.attach_tag(node_id, Constants.TYPE_OF_RESOURCE, tag_id=tag_id)
+
+    def create_node(self, node_config, tags, count) -> Dict[str, Any]:
+        """Creates instances.
+
+        Returns dict mapping instance id to VM object for the created
+        instances.
+        """
+        filters = tags.copy()
+        if TAG_RAY_CLUSTER_NAME not in tags:
+            filters[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+
+        counter = count
+
+        cli_logger.info(f"Create Node tags : {filters}")
+        reused_nodes_dict = {}
+        reuse_nodes = list()
+        reuse_node_ids = []
+        number_of_reused_nodes = 0
+        # Try to reuse previously stopped nodes with compatible configs
+        if self.cache_stopped_nodes:
+            vms = self.vsphere_sdk_client.vcenter.VM.list(
+                VM.FilterSpec(
+                    power_states={
+                        HardPower.State.POWERED_OFF,
+                        HardPower.State.SUSPENDED,
+                    },
+                    clusters={self.cluster_name},
+                )
+            )
+
+            # Select POWERED_OFF or SUSENDED vms which has ray-node-type,
+            # ray-launch-config, ray-user-node-type tags
+            for vm in vms:
+                if number_of_reused_nodes >= counter:
+                    break
+
+                vm_id = vm.name
+                yn_id = DynamicID(type=Constants.TYPE_OF_RESOURCE, id=vm.vm)
+                filter_matched_count = 0
+                for (
+                    tag_id
+                ) in self.vsphere_sdk_client.tagging.TagAssociation.list_attached_tags(
+                    yn_id
+                ):
+                    vsphere_vm_tag = self.vsphere_sdk_client.tagging.Tag.get(
+                        tag_id=tag_id
+                    ).name
+
+                    tag_key_value = vsphere_tag_to_kv_pair(vsphere_vm_tag)
+                    if tag_key_value:
+                        tag_key = tag_key_value[0]
+                        tag_value = tag_key_value[1]
+                        if tag_key in tags and tag_value == filters[tag_key]:
+                            filter_matched_count += 1
```

Comment:
Does it make sense to use `get_matched_tags` for any of this?

Reply:
it makes sense, thanks for helping us find that!